### DISCUSSION TITLE
Trying to get ImageSharp working in the latest Unity3D

### DISCUSSION THREAD

**Discussion by imaginaryhuman** [SixLabors/ImageSharp#1822] [Q&A]
Hey folks. So ImageSharp could potentially fill a lot of useful duties for my app, thank you for developing it. I'm working in Unity, latest version.  So I presume what I'm supposed to do is just dump the source code folder into a unity project, and then unity compiles it right?

When I do this, I get a LOT of compilation errors. I tried both mono and IL2CPP. I get many errors like:

The name "Unsafe" does not exist in the current context
The name "ThrowHelper" does not exist in the current context
"Guard" does not contain a definition for "NotNull"
"Guard" does not contain a definition for "MustBeGreaterThanOrEqualTo"
"Guard" does not contain a definition for "MustBeLessThan"

I switched on 'allow unsafe code' in Unity, didn't seem to make any difference.
There's apparently more than 1000 of these errors.

So.... something is not quite right. Am I missing something? It sounded from a couple other threads that some people were successfully using this in unity? How do I go about getting the library to work?

Thanks

**Comment by tocsoft** [comment]
## Helpers and classes missing
The issue with the Guard class will be because we have some shared infrastructure/code referenced from via Git Submodule that you will also need.

Specifically you should follow the [Manual Build](https://github.com/SixLabors/ImageSharp#manual-build) section of the readme for instructions on pulling down the full set of source required to compile.

## Referencing the library

I believe people will be referencing our published nuget package rather than the source code directly so the compilation issues with missing code wouldn't have been an issue.

*Please note I'm not a unity dev, I have never used it and I believe no one on the core team is a user either.* 

## Realtime usage/Perf
*This is a big cavate on using it from unity.*

`ImageSharp` perf is heavily influenced by the runtime and can be significantly slower when run with different runtimes (our specific target with regards any real perf work is CoreCLR on a machine with SIMD support), the big example I've seen is using it from Blazor, it works but due to the lack of Hardware support for SIMD it is a lot slower, or runtimes with some of the CoreCLR JIT optimizations disabled too.

So please keep in mind `ImageSharp` isn't a library that been designed specifically with the sort of perf trade offs required for real time/per frame rendering, we are probably placed better at generating/loading textures etc up front and once that are then reused rather than continually generated. 

*Please take all that that with a pinch of salt because, as stated earlier, I am no Unity dev and others more familiar with the environment might have different advice and are probably right 😀*

**Comment by imaginaryhuman** [comment]
Thanks. It sounds complicated. I shudder everytime I see anything to do with the 'terminal' being a necessity. Whatever happened to ease of use and just having a zip with a full working version that someone can just load and use right away? Or even better, prebuilt binaries or something. I am somewhat familiar with C# but I really have no idea how to proceed with this on the basis of what's written in your post or on the pages you linked to.

**Reply by JimBobSquarePants** [reply]
>Whatever happened to ease of use and just having a zip with a full working version that someone can just load and use right away? Or even better, prebuilt binaries or something. 

Prebuilt binaries. 

https://www.nuget.org/packages/SixLabors.ImageSharp/

**Comment by imaginaryhuman** [comment]
I see a bunch of 'packages'. Seems the expectation is that someone knows what to do with them and to go into a terminal and run some commands to 'install' them somewhere. I'm not really trying to do that though. I need to get the binaries (dll's etc) into unity.

**Comment by imaginaryhuman** [comment]
I also see a dll. Doesn't work on a mac.

**Comment by imaginaryhuman** [comment]
Does anyone have a full set of current binaries like for windows and for mac intel + apple silicon?

**Reply by JimBobSquarePants** [reply]
Did about 5 seconds googling for you and found this.
https://stackoverflow.com/questions/53447595/nuget-packages-in-unity
 
Unity appears to use it's only package format and requires installing into a specific location. One of the answers on that page led me to this package manager for Unity that you can use to import NuGet packages.
https://github.com/GlitchEnzo/NuGetForUnity

Your working within a proprietary platform that works very differently from the rest of the dotnet ecosystem. I suggest that further questions are asks within the Unity forums.

**Comment by imaginaryhuman** [comment]
Hmm. Well with the unity nugget package importer. .... it imports imagesharp .. but... then this error:

"ArgumentException: The Assembly System.Runtime.CompilerServices.Unsafe is referenced by SixLabors.ImageSharp ('Assets/Packages/SixLabors.ImageSharp.1.0.4/lib/netstandard2.0/SixLabors.ImageSharp.dll'). But the dll is not allowed to be included or could not be found."

By dll maybe it is referring to a binary that doesn't exist for my platform ? I dunno.

**Reply by antonfirsov** [reply]
Looks like the package importer somehow missed that dependency. You can try adding that package too:

https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/

>  I need to get the binaries (dll's etc) into unity.

The package files are basically ZIP files, you can go and grab the netstandard2.0 dll-s from them if you want to do things manually.

Unrelated to the original question, but also make sure you use you are using IL2CPP as scripting backend, otherwise you will face serious perf issues. See #1754.

**Reply by imaginaryhuman** [reply]
Thanks I will try that shortly.

Btw is there plan to provide a dll also that's compatible with Apple Silicon at some point?

**Comment by imaginaryhuman** [comment]
It at least didn't have any other errors besides this one as yet.

**Comment by imaginaryhuman** [comment]
I managed to get it working in unity. Using the unity nuget importer I imported the ImageSharp package from nuget. Then also (eventually) was able to add the missing Unsafe package. 

Then I was able to reference it from script, loaded an image and saved it in another format. :-)

My question now is .... this works on intel Mac x64 via a prebuilt binary/dll. Presumable would work on windows 64-bit too. But. ... what I'd like to happen is to get the source code for the library compiling within Unity/IL2CPP, so that Unity can deal with building for Apple Silicon and other platforms. Do you think that's feasible or likely work, or how I might go about it?

**Reply by tocsoft** [reply]
ImageSharp is completely managed code, there are no native dependencies that are required to be recompiled for different processor architectures. We are only dependent on the .net runtime being made available for the runtime first (which is a Unity/Microsoft level thing).

Quite literally the whole point of this library is to provide a fully featured 2d graphics library that does not need to be recompiled for each processor, our dlls, distributed via nuget, work on windows, linux, and mac... we can even run in the browser via the .net wasm runtime. 

Dlls produced by .net are not native machine code like you would get out of compiling c++, they are much more akin to jar files in the  java ecosystem (as .NET is a virtual machine runtime just like java). Basically the Unity compiler/runtime will be responsible for making it work on those platforms not us.

**Reply by imaginaryhuman** [reply]
So the dll doesn't contain any machine code, it's just like byte code or something, which unity IL2CPP will 'compile' on its own? Or like if running in the unity editor, it's running the bytecode on the .net platform virtual machine?

So then you're saying that all I need is this one dll and it will work "on .net" as a platform via JIT, and if used in Unity via AOT IL2CPP it will be turned into c++ and compiled to native machine code, right?

What about the extra little dll I had to install for the missing dependency on the 'Unsafe' library?

**Reply by tocsoft** [reply]
> So the dll doesn't contain any machine code, it's just like byte code or something, which unity IL2CPP will 'compile' on its own? Or like if running in the unity editor, it's running the bytecode on the .net platform virtual machine? 

Yep that's exactly how it'll be working. (compilation level anyway not sure on the editor runtime side, but your guess sounds like its probably correct)

> So then you're saying that all I need is this one dll and it will work "on .net" as a platform via JIT, and if used in Unity via AOT IL2CPP it will be turned into c++ and compiled to native machine code, right?

Yep, that's how its supposed to work as far as I know.

> What about the extra little dll I had to install for the missing dependency on the 'Unsafe' library?

I would be highly surprised if it also didn't just work for that one too, but ultimately I feel your just going to have to try it on the various platforms your trying to target and verify it for yourself (I personally wouldn't want to bet on any 3rd party dependency that I wasn't sure would work on my target platforms without verifying it for myself first).

**Reply by imaginaryhuman** [reply]
Good idea, thanks.

**Reply by imaginaryhuman** [reply]
Suppose that means I have to buy a new Mac, lol
